src/RunAction.cpp

```diff
@@ -220,11 +224,26 @@ NodeBuildResult::Enum RunAction(BuildQueue *queue, ThreadState *thread_state, Ru
 
         bool allowUnwrittenOutputFiles = (node_data->m_Flags & Frozen::DagNode::kFlagAllowUnwrittenOutputFiles);
         if (!allowUnwrittenOutputFiles)
+        {
+            uint64_t current_time = time(NULL);
+
             for (int i = 0; i < n_outputs; i++)
             {
                 FileInfo info = GetFileInfo(node_data->m_OutputFiles[i].m_Filename);
                 pre_timestamps[i] = info.m_Timestamp;
+
+                if (info.m_Timestamp == current_time)
+                {
+                    // This file has been created so recently that a very fast action might not
+                    // actually be recognised as modifying the timestamp on the file. To avoid
+                    // this, backdate the file by a second, so that any actual activity will definitely
+                    // cause the timestamp to change.
+                    struct utimbuf times;
+                    pre_timestamps[i] = times.actime = times.modtime = current_time - 1;
+                    utime(node_data->m_OutputFiles[i].m_Filename, &times);
```

Comment:
Could this cause unneccesary rebuilds for other actions that depend on one of these backdated files?

Reply:
These files are (supposedly) about to get overwritten with new content/timestamps, in the action that we are about to run - so anything that is using them as an input is _supposed_ to get rebuilt after this. The only situation where this will change the timestamp that would not have previously changed the timestamp is if this node fails to run properly, in which case we will be failing the action and halting the build before running any actions that are dependent on this.